src/tss2-esys/esys_iutil.c

```diff
@@ -631,50 +664,58 @@ iesys_decrypt_param(ESYS_CONTEXT * esys_context,
+        if (rsrc_session->type_policy_session == POLICY_PASSWORD) {
+            /* A policy password session has not auth value */
+            if (rspAuths->auths[i].hmac.size != 0) {
+                LOG_ERROR("PolicyPassword session's HMAC must be 0-length.");
+                return TSS2_ESYS_RC_RSP_AUTH_FAILED;
             }
-            TPM2B_AUTH rp_hmac;
-            rp_hmac.size = sizeof(TPMU_HA);
-            rsrc_session->nonceTPM = rspAuths->auths[i].nonce;
-            rsrc_session->sessionAttributes =
-                rspAuths->auths[i].sessionAttributes;
-            // TODO check: auths.auths[i].hmac.size =  sizeof(TPMU_HA);
-            TSS2_RC r =
-                iesys_crypto_authHmac(rsrc_session->authHash,
-                                      &rsrc_session->sessionValue[0],
-                                      rsrc_session->sizeSessionValue,
-                                      &rp_hash_tab[hi].digest[0],
-                                      rp_hash_tab[hi].size,
-                                      &rsrc_session->nonceTPM,
-                                      &rsrc_session->nonceCaller, NULL, NULL,
-                                      rspAuths->auths[i].sessionAttributes,
-                                      &rp_hmac);
-            return_if_error(r, "HMAC error");
-            if (!cmp_TPM2B_AUTH(&rspAuths->auths[i].hmac, &rp_hmac)) {
-                LOG_ERROR("Error: Invalid hmac response.");
-                return TSS2_ESYS_RC_BAD_VALUE;
+            continue;
+        }
+
+        /* Find the rpHash for the hash algorithm used by this session */
+        int hi;
+        for (hi = 0; hi < rpHashNum; hi++) {
+            if (rsrc_session->authHash == rp_hash_tab[hi].alg) {
+                break;
             }
         }
+        if (hi == rpHashNum) {
+            LOG_ERROR("rpHash for alg %"PRIx16 " not found.",
+                      rsrc_session->authHash);
+            return TSS2_ESYS_RC_GENERAL_FAILURE;
+        }
 
+        TPM2B_AUTH rp_hmac;
+        rp_hmac.size = sizeof(TPMU_HA);
+        rsrc_session->nonceTPM = rspAuths->auths[i].nonce;
+        rsrc_session->sessionAttributes =
+            rspAuths->auths[i].sessionAttributes;
+        r = iesys_crypto_authHmac(rsrc_session->authHash,
+                                  &rsrc_session->sessionValue[0],
+                                  rsrc_session->sizeSessionValue,
+                                  &rp_hash_tab[hi].digest[0],
+                                  rp_hash_tab[hi].size,
+                                  &rsrc_session->nonceTPM,
+                                  &rsrc_session->nonceCaller, NULL, NULL,
+                                  rspAuths->auths[i].sessionAttributes,
+                                  &rp_hmac);
+        return_if_error(r, "HMAC error");
+
+        if (!cmp_TPM2B_AUTH(&rspAuths->auths[i].hmac, &rp_hmac)) {
+            LOG_ERROR("Invalid in TPM's response for session %i", i);
```

Comment:
s/Invalid in TPM's/Invalid TPM/

or you can word it as:
"TPM's response is invalid for session %i"

Reply:
done